### PR TITLE
feat(root) implement external signer for mpcv2

### PR DESCRIPTION
**Pull request by alebusse**
added base utils for MPCv2 external signer
added MPCv2 base code to support external signer
Implemented singing using external signer for MPCv2
Use the hashed signableTx as Adata to verify that the encrypted session matches with the tx during
the second and third DSG rounds in external signer, this would prevent re using the stored session
for a different tx

WP-2136

### CODE REVIEW

File: modules/express/src/clientRoutes.ts [hunk 1]
```diff
@@ -828,7 +855,7 @@ function createSendParams(req: express.Request) {
   }
 }
 
-function createTSSSendParams(req: express.Request) {
+function createTSSSendParams(req: express.Request, wallet?: Wallet) {
```
Comment (zahin-mohammad) [inline modules/express/src/clientRoutes.ts:858]: shouldn't wallet be required? Are there routes that aren't passing it in?

File: modules/sdk-core/src/bitgo/utils/tss/baseTypes.ts [hunk 2]
```diff
@@ -107,6 +110,39 @@ export interface CustomGShareGeneratingFunction {
     bitgoToUserCommitment: CommitmentShareRecord;
   }): Promise<GShare>;
 }
+
+export interface CustomMPCv2Round1GeneratingFunction {
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/utils/tss/baseTypes.ts:114]: Nit: signing and key gen both have multiple rounds, would be clearer if we specified signing here.
```suggestion
export interface CustomMPCv2SigningRound1GeneratingFunction {
```

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts [hunk 3]
```diff
@@ -724,5 +718,306 @@ export class EcdsaMPCv2Utils extends BaseEcdsaUtils {
...
+    externalSignerMPCv2Round3Generator: CustomMPCv2Round3GeneratingFunction,
+    requestType: RequestType = RequestType.tx
+  ): Promise<TxRequest> {
+    const { txRequest, reqId } = params;
+    let txRequestResolved: TxRequest;
+
+    assert(
+      requestType === RequestType.tx,
+      'Only transaction signing is supported for external signer, got: ' + requestType
+    );
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:801]: Lets add a ticket to support message signing with external signer. Don't want to let it fall through the cracks even though it is not needed atm.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 4]
```diff
@@ -3152,6 +3161,54 @@ export class Wallet implements IWallet {
...
+    if (!params.customMPCv2Round3GenerationFunction) {
+      throw new Error('Generator function for MPCv2 Round 3 share required to sign transactions with External Signer.');
+    }
+
+    try {
+      assert(this.tssUtils, 'tssUtils must be defined');
+      const signedTxRequest = await this.tssUtils.signEcdsaMPCv2TssUsingExternalSigner(
+        {
+          txRequest: txRequestId,
+          prv: '',
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:3199]: Nit: we should allow undefined `prv` if the custom signing functions are passed. Seems weird to pass an empty string.

File: modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts [hunk 5]
```diff
@@ -154,6 +157,26 @@ export default class BaseTssUtils<KeyShare> extends MpcUtils implements ITssUtil
...
+   * @returns {Promise<TxRequest>} - a signed tx request
+   */
+  signEcdsaMPCv2TssUsingExternalSigner(
+    params: TSSParams | TSSParamsForMessage,
+    externalSignerMPCv2Round1Generator: CustomMPCv2Round1GeneratingFunction,
+    externalSignerMPCv2Round2Generator: CustomMPCv2Round2GeneratingFunction,
+    externalSignerMPCv2Round3Generator: CustomMPCv2Round3GeneratingFunction,
+    requestType?: RequestType
+  ): Promise<TxRequest> {
+    throw new Error('Method not implemented.');
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:177]: This still needs to be implemented in this PR right? Or will this be addressed in a different PR?

Comment (alebusse) [reply modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:177]: Nop, the base utils use the non implemented errors incase you used the wrong util (like eddsa)

File: modules/express/src/clientRoutes.ts (hunk 1 above)
Comment (alebusse) [reply modules/express/src/clientRoutes.ts:858]: its not required for Eddsa since we dont check the multisigTypeVersion, but i can make it required

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 4 above)
Comment (alebusse) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:3199]: Ya, its mostly because its using the wrong type. i will fix it

**Review by zahin-mohammad — DISMISSED** [review]

**Review by zahin-mohammad — APPROVED** [review]

**Review by islamaminBitGo — APPROVED** [review]
